openff/system/components/foyer.py

```diff
@@ -0,0 +1,152 @@
+    return list(filter(lambda x: isinstance(x, gen_type), ff.getGenerators())).pop()
+
+
+def _topology_from_parmed(structure: pmd.Structure) -> Topology:
+    """ToDo: Create a openFF Topology from a parmed structure"""
+
+
+def from_foyer(structure: pmd.Structure, ff: Forcefield, **kwargs) -> System:
+    """Create an openFF system object from a parmed structure by applying a foyer Forcefield"""
+    system = System()
+    system.topology = _topology_from_parmed(structure)
+    foyer_atom_handler = FoyerAtomTypes()
+    foyer_bond_handler = FoyerBondHandler()
+    system.handlers["FoyerAtomHandler"] = foyer_atom_handler
+    system.handlers["FoyerBondHandler"] = foyer_bond_handler
+
+    return system
+
+
+class FoyerAtomTypes(PotentialHandler):
+
+    name: str = "Atoms"
+    expression: str = "4*epsilon*((sigma/r)**12-(sigma/r)**6)"
+    independent_variables: Set[str] = {"r"}
+    slot_map: Dict[int, str] = dict()  # type: ignore
+    potentials: Dict[str, Potential] = dict()
+
+    def store_matches(
+        self,
+        forcefield: Forcefield,
+        structure: pmd.Structure,
+    ) -> None:
+        """
+        Populate self.slot_map with key-val pairs of slots
+        and unique potential identifiers
+
+        """
+        type_map = forcefield.run_atomtyping(structure)
+        for key, val in type_map.items():
+            self.slot_map[key] = val["atomtype"]
+
+    def store_potentials(self, forcefield) -> None:
+        from simtk.openmm.app.forcefield import NonbondedGenerator  # type: ignore
+
+        non_bonded_forces_gen = _get_openmm_force_gen(forcefield, NonbondedGenerator)
+
+        if non_bonded_forces_gen:
+            nonbonded_params = non_bonded_forces_gen.params.paramsForType
+
+            for atom_idx in self.slot_map:
+                try:
+                    params = _copy_params(
+                        nonbonded_params[self.slot_map[atom_idx]],
+                        "charge",
+                        param_units={"epsilon": u.kcal / u.mol, "sigma": u.nm},
+                    )
+                except KeyError:
+                    raise MissingParametersError(
+                        f"Missing parameters for atomtype {self.slot_map[atom_idx]}"
+                        "in the forcefield"
+                    )
+
+                self.potentials[self.slot_map[atom_idx]] = Potential(parameters=params)
+        else:
+            raise MissingParametersError(
+                "The forcefield is missing NonBondedForces Parameters"
+            )
+
+
+class FoyerBondHandler(PotentialHandler):
+
+    name: str = "Bonds"
+    expression: str = "1/2 * k * (r - length) ** 2"
+    independent_variables: Set[str] = {"r"}
+    slot_map: Dict[Tuple[int, int], Tuple[str, str]] = dict()  # type: ignore
+    potentials: Dict[Tuple[int, int], Potential] = dict()  # type: ignore
+
+    def store_matches(
+        self, structure: pmd.Structure, atom_slots: Dict[str, str]
+    ) -> None:
```

Comment:
Since Foyer types bonds (and other sections) using the same typing algorithm as the non-bonded section, it would be better to use those results here and **not** look into a ParmEd structure 

```suggestion
    def store_matches(
        self, atom_types: FoyerAtomTypes,
    ) -> None:
```

Reply:
I was under the impression that the `run_atomtyping` step doesn't parameterize the structure. The tricky thing here is to get the types for underlying bonds in the structure/ compound. Am I missing something?

Reply:
I had to look this up - I thought that `foyer.Forcefield.run_atomtyping` took in an un-parametrized ParmEd structure  and returned the same object with atom types stored as string on each atom. But I was wrong - iit does give something more compact and useful as an output:


```
foyer.Forcefield(name='oplsaa').run_atomtyping(compound.to_parmed())
/Users/mwt/miniconda3/envs/openff-system/lib/python3.8/site-packages/foyer/forcefield.py:448: UserWarning: No force field version number found in force field XML file.
  warnings.warn(
/Users/mwt/miniconda3/envs/openff-system/lib/python3.8/site-packages/foyer/forcefield.py:460: UserWarning: No force field name found in force field XML file.
  warnings.warn(
/Users/mwt/miniconda3/envs/openff-system/lib/python3.8/site-packages/foyer/validator.py:132: ValidationWarning: You have empty smart definition(s)
  warn("You have empty smart definition(s)", ValidationWarning)
Out[13]:
{0: {'whitelist': set(), 'blacklist': set(), 'atomtype': 'opls_135'},
 1: {'whitelist': set(), 'blacklist': set(), 'atomtype': 'opls_157'},
 2: {'whitelist': set(), 'blacklist': set(), 'atomtype': 'opls_154'},
 3: {'whitelist': set(), 'blacklist': set(), 'atomtype': 'opls_140'},
 4: {'whitelist': set(), 'blacklist': set(), 'atomtype': 'opls_140'},
 5: {'whitelist': set(), 'blacklist': set(), 'atomtype': 'opls_140'},
 6: {'whitelist': set(), 'blacklist': set(), 'atomtype': 'opls_140'},
 7: {'whitelist': set(), 'blacklist': set(), 'atomtype': 'opls_140'},
 8: {'whitelist': set(), 'blacklist': set(), 'atomtype': 'opls_155'}}
```

It would still, in my opinion, be useful to have that accept a compound as an argument, more closely matching one of the intended pathways for data to flow through foyer (`Forcefield.apply(compound)`). The data that Foyer actually needs from the un-parametrized ParmEd structure is fairly minimal (elements, bonds, residues, maybe formal charge) compared to all of the data that can be shoved into a structure. An OpenFF Topology object could serve this purpose well - probably out of the box but possibly with a thin subclass.